src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -359,6 +359,76 @@ def _encode_extended_key_usage(backend, extended_key_usage):
     return pp, r
 
 
+_CRLREASONFLAGS = {
+    x509.ReasonFlags.key_compromise: 1,
+    x509.ReasonFlags.ca_compromise: 2,
+    x509.ReasonFlags.affiliation_changed: 3,
+    x509.ReasonFlags.superseded: 4,
+    x509.ReasonFlags.cessation_of_operation: 5,
+    x509.ReasonFlags.certificate_hold: 6,
+    x509.ReasonFlags.privilege_withdrawn: 7,
+    x509.ReasonFlags.aa_compromise: 8,
+}
+
+
+def _encode_crl_distribution_points(backend, crl_distribution_points):
+    cdp = backend._lib.sk_DIST_POINT_new_null()
+    cdp = backend._ffi.gc(cdp, backend._lib.sk_DIST_POINT_free)
+    for point in crl_distribution_points:
+        dp = backend._lib.DIST_POINT_new()
+        assert dp != backend._ffi.NULL
+
+        if point.reasons:
+            bitmask = backend._lib.ASN1_BIT_STRING_new()
+            assert bitmask != backend._ffi.NULL
+            dp.reasons = bitmask
+            for reason in point.reasons:
+                res = backend._lib.ASN1_BIT_STRING_set_bit(
+                    bitmask, _CRLREASONFLAGS[reason], 1
+                )
+                assert res == 1
+
+        if point.full_name:
+            dpn = backend._lib.DIST_POINT_NAME_new()
+            assert dpn != backend._ffi.NULL
+            # Type 0 is fullName, there is no #define for it in the code.
+            dpn.type = 0
+            for name in point.full_name:
+                gns = backend._lib.GENERAL_NAMES_new()
+                assert gns != backend._ffi.NULL
+                for name in point.full_name:
+                    gn = _encode_general_name(backend, name)
+                    res = backend._lib.sk_GENERAL_NAME_push(gns, gn)
+                    assert res >= 1
+
+            dpn.name.fullname = gns
```

Comment:
This doesn't look right, `gns` is created inside the `for name in` loop, but assigned outside. This means you'd only ever at the last one used.